Code review of entire module for transfer to Rally

jeffmay opened the pull request:
@rallyhealth/Engineers @arkban @usufruct99 @slustbader @galak-fyyar @rh-nalitkin

This is every commit from the beginning of time in a PR, so that comments can be made. I will address any comments in a separate PR.

Please review the code to make sure it meets Rally Health's code quality standards.

Note: The license is Apache 2.0, but it might need to change to MIT. This is not visible in this PR, because the [LICENSE](https://github.com/jeffmay/play-json-ops/blob/master/LICENSE) file was added in the first commit to the repo.

Depends on changes:
- [x] https://github.com/rallyhealth/play-json-ops/pull/22 - Change of license and published artifact location

Commit:
Initial commit

- Added JsonImplicits
- Added TupleFormats
- Added DurationFormat
- Added FormatOps
- Added JsResults

Commit:
Ignore .idea and add build.properties.

Commit:
Added initial set of helper classes.

Squashed from multiple commits:
- Added JsonImplicits
- Added AbstractJsonOps
- Added JsonTransform for implicitly redacting json data
- Added ScalaCheck and JsValueGenerators
- Added JsonTransform
- Added JsValueGenerators with Width and Height.

Commit:
Added UTCFormats

Commit:
bump version 0.2.0

Commit:
Bump to 0.2.1

- Added README.md
- Updated play-json version to 2.3.7
- Cross-compiles to Scala 2.11.6
- Arbitrary[JsValue] now accepts implicit Depth
- Updated specs to be more robust

Commit:
Fixed bugs in DurationFormat

- Added DurationOps
- Added DurationGenerators
- Added DurationGeneratorsSpec
- Added DurationFormatSpec
- Added new play.api.libs.json.ops._ style import

Commit:
Added dependency to scalacheck-ops

- Added JsonFormatSpec

Commit:
Updates to the README.md

Commit:
Moved some code around for consistent style.

- Deprecated import of JsonImplicits in favor of importing play.api.libs.json.ops._

Commit:
Added framework agnostic automatic SerializationTests

- Added GenericTestSuite and ScalaTestBridge
- Added SerializationTests for automatic framework agnostic
  Serialization unit tests
- Added ScalaTestBridge for running the automatic tests in ScalaTest
- Added PlayJsonFormatSpec for play / scalatest specific unit tests

Commit:
Added ScalaTest and ScalaCheck to compile scope dependencies

Commit:
Switched DurationFormatSpec to generated SerializationTests

Commit:
Removed ParallelTestExecution to avoid ugly test output.

Commit:
Added JsValue shrinks

- Added support for Shrink in PlayJsonFormatSpec

jeffmay committed:
Merge pull request #1 from jeffmay/finite-duration

Version 0.2.2

Commit:
Fixed compiler errors with 2.11.6 compilation.

- Removed unnecessary commented code

Commit:
Bump version to v0.2.2

Commit:
Fixed bug with mixing ForgivingDurationFormat with the others.

Commit:
Updated SerializationTests to be less Play specific.

- Updated comments from v0.2.2

Commit:
0.2.4 - Updated docs and publish to bintray

Commit:
1.0.0 - Release version

- Added ScalaDoc links favoring source links over ScalaDoc
- Updated build.sbt for safer compilation

Commit:
Minor sbt updates

- disable publishing of ScalaDoc and JavaDoc
- added deprecation flag for 2.10

Commit:
1.1.0 - Split out test code helpers into play-json-tests

- No longer publishing internal test code
- Moved DurationOps code into play-json-tests lib

Commit:
Fixed issue with fail method not being called appropriately.

Commit:
Fixed build.sbt issues around publishing.

Commit:
1.1.1 - Fixed issues with build and minor bug

Commit:
Added .travis.yml to the build.

Commit:
Added resolvers.

jeffmay committed:
Merge pull request #2 from jeffmay/travis

Added .travis.yml to the build.

Commit:
Fixed strange build dependency issue.

Commit:
Fixed symmetric serialization of JsValue not equaling itself.

- Updated tests to verify this property of JsValueGenerators
- Fixed up the generators to generate safer numbers and avoid
  JsUndefined in nested arrays, objects, and non-primitive JsValue
  generators.

jeffmay committed:
Merge pull request #3 from jeffmay/fix-generators

Fixed symmetric serialization of JsValue not equaling itself.

Commit:
Added coveralls and updated travis build.

jeffmay committed:
Merge pull request #6 from jeffmay/coveralls

Added coveralls and updated travis build.

jeffmay committed:
Update README.md

Added travis, bintray, and coveralls badges.

jeffmay committed:
Update README.md

Added play-json-tests bintray badge

Commit:
[NOJIRA] - make playJsonOps not dependent on scalacheck and scalatest

jeffmay committed:
Merge pull request #7 from vodopis/test-dependency-fix

[NOJIRA] - make playJsonOps not dependent on scalacheck and scalatest

Commit:
Bump version to 1.1.2

Commit:
Bump version to minor change since dependency tree changed.

Commit:
Fixes #8

- Add OFormatPure to OFormatOps.

Commit:
Update to latest supported 2.10 versions.

Commit:
Updated AbstractJsonOps support objects and better exceptions

- Updated AbstractJsonOps to support object serialization
- Improved syntax for formatWithType to utilize inference
- Deprecated old formatWithType
- Added exception to handle mismatched formatAbstract mappings
- Added Json.extractTypeKey[X].usingKeyObject()
- Added more tests

Commit:
Updated build settings.

- Updated build.sbt to use explicit root project.
- Upgrade to Scala 2.11.7
- Upgrade sbt-bintray
- Upgrade scalacheck-ops
- Updated travis.yml to use cache and fix coverageAggregate.

jeffmay committed:
Merge pull request #9 from jeffmay/updates

#8 and updated versions

Commit:
Bump version to 1.2.1

Commit:
Updated tests to use assertResult for better error messaging.

jeffmay committed:
Update LICENSE

Updated copyright date to match repo creation year.

jeffmay committed:
Merge pull request #10 from jeffmay/final-1.2.x-updates

Final 1.x updates

Commit:
Bump version to 1.3.0

jeffmay committed:
Upgraded scalacheck-ops to 1.3.0 and build plugins

jeffmay committed:
Merge pull request #13 from jeffmay/upgrade-scalacheck-ops

Upgraded scalacheck-ops to 1.3.0 and build plugins

jeffmay committed:
Bump version to 1.3.1

jeffmay committed:
Moved test code out of play-json-ops to avoid dependency on test project.

jeffmay committed:
Bump version to 1.4.0.

jeffmay committed:
Merge pull request #14 from jeffmay/fix-dependency-graph

Moved test code out of play-json-ops to avoid dependency on test project

jeffmay committed:
Copied projects into Play 2.5.10 compatible versions.

- Upgraded sbt
- Added coursier

jeffmay committed:
Merge pull request #15 from jeffmay/play25

Copied projects into Play 2.5.10 compatible versions.

jeffmay committed:
Moved code into the correct directories.

jeffmay committed:
Merge pull request #16 from jeffmay/cleanup-upgrade

Moved code into the correct directories.

jeffmay committed:
Update README.md

jeffmay committed:
Support cross-compilation for ScalaTest for 2.0.0 release

- Rename artifacts to be more consistent with other cross-compiled
  projects

jeffmay committed:
Support legacy artifact names to release minor upgrade.

jeffmay committed:
Removed deprecated methods for v2 release

- Renamed artifacts
- Minor cleanup

jeffmay committed:
Update README.md

- Add new artifacts
- Reformatted artifact list
- Fixed documentation regarding extending JsonImplicits
- Added missing docs for OFormat.pure

jeffmay committed:
Updated README.md to add v2 branch artifact names

jeffmay committed:
Merge pull request #17 from jeffmay/scalacheck-13

Upgrade to support cross-compiled scalacheck 2.2.6 and 3.0.4

jeffmay committed:
Fixed broken artifact names

- Remove cross compilation

jeffmay committed:
Merge pull request #18 from jeffmay/artifact-name-fixes

Fixed broken artifact names

jeffmay committed:
Fix README to use non-deprecated methods for abstract formats

jeffmay committed:
Added abstractFormatOps for binary compatibility

- Ignore deprecated method

jeffmay committed:
Merge pull request #20 from jeffmay/bincompat

Added abstractFormatOps for binary compatibility

jeffmay committed:
Bump version to 1.6.1

jeffmay committed:
Merge pull request #19 from jeffmay/v2

Remove deprecated methods and fix artifact names for 2.0.0 release

arkban commented:
Um, this is on the "jeffmay" org, shouldn't it be on the "Rallyhealth" org?

arkban commented:
> Note: The license is Apache 2.0, but it might need to change to MIT. 

I thought APL was okay.

arkban commented on .gitignore:
```diff
@@ -12,6 +12,9 @@ src_managed/
 project/boot/
 project/plugins/project/
 
+# IntelliJ Specific
```
I've heard good arguments that this should be in users global `.gitignore`

jeffmay replied:
Probably... but I can't imagine anyone not using IntelliJ for Scala 😲

arkban replied:
Sounds like a challenge for Vito ;)

arkban commented on .travis.yml:
```diff
@@ -0,0 +1,22 @@
+language: scala
```
If this is going to Rally I think we're going to want to use CJE instead of travis or anything else.

jeffmay replied:
We are currently using travis for all of our open-source Scala projects. We can discuss changing this in a separate thread.

arkban replied:
TIL, no need to discuss if it's the way we do things for our OSS projects

arkban commented on LICENSE:
```diff
@@ -186,7 +186,7 @@ Apache License
       identification within third-party archives.
 
-   Copyright {yyyy} {name of copyright owner}
+   Copyright 2015 Jeff May
```
Um...

jeffmay replied:
I can update this before transferring ownership of the repo.

arkban commented on README.md:
```diff
@@ -1,2 +1,372 @@
+<h1>Versions</h1>
+
+<h2>2.X Branch</h2>
+<table>
```
You know you can write tables in markdown, right?

jeffmay replied:
I remember when I first attempted to do it in markdown, it wasn't powerful enough to do what I wanted, so I used HTML. I can give it another try. Maybe a fast-follow item?

arkban replied:
It's fine, just letting you know, doesn't need to be a fast follow, sounds more like an enhancement that will come when someone gets fed up modifying the HTML table ;p

arkban commented on build.sbt:
```diff
@@ -0,0 +1,96 @@
+
+    resolvers ++= Seq(
+      "Typesafe Releases" at "http://repo.typesafe.com/typesafe/releases/",
+      "jeffmay at bintray" at "https://dl.bintray.com/jeffmay/maven"
```
Um...

jeffmay replied:
This review is about transferring ownership to Rally. It isn't published to Rally currently. In fact, this PR isn't about merging anything. It's just about reviewing the code. I can see the confusion though. How about I transfer the repo to Rally Health and then do a review?

arkban replied:
Either that or just clarify, because I (clearly) didn't understand that.

arkban commented on build.sbt:
```diff
@@ -0,0 +1,96 @@
+name := "play-json-ops-root"
+organization in ThisBuild := "me.jeffmay"
```
I'm not sure the details of the transfer, but if you are giving ownership to Rally this should probably be Rally

arkban commented on project/Dependencies.scala:
```diff
@@ -0,0 +1,28 @@
+      case `scalatest2Version` => ""
+      case `scalatest3Version` => "_1-13"
+    }
+    "me.jeffmay" %% s"scalacheck-ops$suffix" % scalacheckOpsVersion
```
Is this correct?

arkban reviewed (changes requested):
I reviewed the files in the root directory, but that's it, not the code. If I get time I can try to do that later.

jeffmay committed:
Change of software license to MIT and publish under Rally Health

rcmurphy committed:
Merge pull request #22 from rallyhealth/move-to-rally

Change of software license to MIT and publish under Rally Health

jeffmay committed:
Publish and resolve with maven respository

jeffmay committed:
Update README.md to with corrected links

jeffmay committed:
Merge pull request #23 from rallyhealth/publish-to-maven

Publish and resolve with maven respository

jeffmay committed:
Fix publish to bintray urls

jeffmay committed:
Remove hard-coded version

jeffmay committed:
Invalidate implicit resolution for JsonImplicits.formatMap

- This removes an unsafe implicit landmine that would call .toString
  as a default implicit argument
- These changes are still binary compatible
- Add unit tests for ambiguous implicit fix

jeffmay committed:
Update build to use codecov.io instead of coveralls

jeffmay committed:
Update README.md to use new badges

jeffmay committed:
Update README.md

Fix badge links for Bintray latest versions
